### ISSUE TITLE
Issues with cloning of aligner

### ISSUE THREAD

**Issue by rob-p** [jguhlin/minimap2-rs#71]
Hi @jguhlin,

  First; thanks again for the great work on this. I'm already finding minimap2-rs *super* useful, and we're looking to use it to include a "raw read" mode in our `oarfish` tool.  I noticed two specific issues with cloning an aligner and wondered if you had any thoughts or insight into them.

  So I'm following the basic strategy employed in fake minimap2 / [minimap2-rs-test](https://t.co/93KFQpwLjW), where I have several threads that take reads (in my case over a channel) and then send the reads to another consumer.  Each thread needs it's own aligner object and I am calling `clone()` to get it.

  I'm having two problems.  First, for context, since I need references to local variables, I am using a `std::thread::scope` and scoped threads which may be why I see this clearly in my example but the other examples I'm building off of don't.

  1) Since each clone gets it's own pointer to the index (the `idx` field), they all try to `drop` the index (which calls `free` in the FFI) when the thread is joined and the scope ends.  However, this obviously leads to a double-free and, in my case, a segfault.  I can get around this by explicitly setting the `idx` field of each cloned `Aligner` to `None` before the thread returns, but this feels error prone.  I wonder if it would make sense to have either (a) some properly resource-managed pointer type for the index (e.g. an `Arc` or some such), or (b) some specific idx "view" type that essentially constitutes a weakptr to the original index.  Either way, it'd be great to have some sort of way to do the clone in a carefree manner but not worry about double-free.

  2) It's not clear to me the "depth" of the clone.  Specifically, while in the original aligner I call the `with_cigar` method on the builder, the clones aren't seeming to provide this information by default and instead I'm having to call `aligner.clone().with_cigar()` to get a clone that also provides CIGAR strings.  I guess the expected cloning behavior is for the clone to have all of the same properties set as the original, and the real fear I have here is that I don't know what other fields might be affected apart from the CIGAR string generation (perhaps it's simply the flags that aren't deep cloned)?

Anyway, I'm happy to provide further detail about the above, or to discuss potential solutions or design decisions about addressing these.  Let me know your thoughts, and thanks again for minimap2-rs!

--Rob

**Comment by jguhlin** [comment]
It's my first FFI project, so has been a learning experience. I like Arc as a solution.

Surprised about the second point, I'll dig into it this week! Clone should produce full clones.

**Comment by Sam-Sims** [comment]
Hi @rob-p hope you don't mind me jumping in,

I am having the same issue you outline in your first point (I am using crossbeam channels in a similar way to the "fake minimap2") - but after cloning the `Aligner` and joining the threads I get a double free segfault.

I tried doing something like `aligner.idx = None;` in the thread after calling `aligner.map` but I still get a segfault - do you have an example of how you implemented threading?

Cheers

**Comment by rob-p** [comment]
Hi @Sam-Sims,

  You can find the relevant line (and closure) in my code here https://github.com/COMBINE-lab/oarfish/blob/5b09b4345b609559a1aa7d76df75bf3aebac6ba9/src/bulk.rs#L334.  Basically, you only let the original aligner retain the pointer and make sure you set it to None before the point of dropping within each thread.  Let me know if this helps.

--Rob

**Comment by Sam-Sims** [comment]
Ah thanks that's super useful! I needed to set it to None in the thread closure. Agree it would be nice to be able to clone an`Aligner` without having to worry about segfault.

**Comment by mbhall88** [comment]
I have something which *seems* to be working well (i.e., I haven't got any seqfaults etc.)

Essentially I created an AlignerWrapper which holds an `Arc` to an `Aligner`, then I [`Arc::clone`](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.clone) the aligner in each spawned thread. Here is some example code

```rust
struct AlignerWrapper {
    /// Shared aligner across threads
    aligner: Arc<Aligner>, 
}

impl AlignerWrapper {
    fn new(target_file: &Path, threads: usize) -> Result<Self, LrgeError> {
        let mut aligner = Aligner::builder()
            .ava_ont()
            .with_index_threads(threads)
            .with_index(target_file, None)
            .unwrap();

        Ok(Self {
            aligner: Arc::new(aligner),
        })
    }

    fn align_reads(&self, query_file: PathBuf) -> Result<Vec<Mapping>, LrgeError> {
        // Bounded channel to control memory usage
        let (sender, receiver) = channel::bounded(1000); 

        // Shared reference for the producer thread
        let aligner = Arc::clone(&self.aligner); 

        // Producer: Read FASTQ records and send them to the channel
        let producer = std::thread::spawn(move || -> Result<(), LrgeError> {
            let mut fastx_reader = parse_fastx_file(query_file).map_err(|e| {
                LrgeError::FastqParseError(format!("Error parsing query FASTQ file: {}", e))
            })?;

            while let Some(record) = fastx_reader.next() {
                match record {
                    Ok(rec) => {
                        let msg = io::Message::Data((
                            rec.read_id().to_owned(),
                            Vec::from(rec.seq().to_owned()),
                        ));
                        if sender.send(msg).is_err() {
                            break; // Exit if the receiver is dropped
                        }
                    }
                    Err(e) => {
                        return Err(LrgeError::FastqParseError(format!(
                            "Error parsing query FASTQ file: {}",
                            e
                        )));
                    }
                }
            }

            // Close the channel to signal that no more records will be sent
            drop(sender);

            Ok(())
        });

        // Consumer: Process records from the channel in parallel
        let alignments: Vec<Mapping> = receiver
            .into_iter()
            .par_bridge()  // gosh rayon is ridiculously ergonomic
            .flat_map(|record| {
                let (rid, seq) = match record {
                    io::Message::Data(data) => data,
                    io::Message::End => unimplemented!("End message not expected here"),
                };
                debug!("Processing read: {:?}", String::from_utf8_lossy(&rid));
                let mut qname = rid.to_owned();
                if !(qname.last() == Some(&0)) {
                    qname.push(0);
                }
                // Use the shared aligner to perform alignment
                aligner.map(&seq, Some(&rid)).unwrap()
            })
            .collect();

        // Wait for the producer to finish
        producer.join().unwrap();

        Ok(alignments)
    }
}
```

Note, this is using a small API wrapper I have written around minimap-sys so some calls to things like `map` are a little different. But hopefully you get the gist.

I'd be interested to know if this works for others or if I am just not seeing the problem pop up yet and need to do more thorough stress testing.

**Comment by rob-p** [comment]
Hi @mbhall88,

  I believe your specific architecture should work, since you have only one underlying `Aligner` object that is accessed through the `Arc`. Rather @Sam-Sims and I were attempting to use the approach shown in the [example application](https://github.com/jianshu93/minimap2-rs-test/blob/main/src/main.rs), where each consumer (i.e. mapping) thread makes it's own clone of the `Aligner` struct.  Granted, I don't know what effect the extra `Arc` indirection will have on performance, but the idea of having a local clone in each thread is so that the `Aligner`'s avoid the extra redirection, and also (perhaps) so they can have some different (non-index-related) settings if one wants.

  If your architecture works well in your use case, I believe it seems sound to me.  However, I still think it would be useful if we could avoid the undefined behavior (i.e. use-after free) that occurs when a cloned `Aligner` is used in different scopes.  There, the end of the first closing scope drops the `Aligner` and destroys the underlying index pointer.

--Rob

**Comment by Sam-Sims** [comment]
I did take a stab at this a while ago, implementing a new struct for the index - `MmIndex`, and then using an `Arc` to manage it

Something like:
```rust
pub struct MmIdx(*mut mm_idx_t);
impl Drop for MmIdx {
    fn drop(&mut self) {
        unsafe { mm_idx_destroy(self.0) };
    }
}
```
```rust
pub struct Aligner {
    ...
    pub idx: Option<Arc<MmIdx>>,
}
```

and then removing the original drop implementation for the aligner. 

I just pushed the changes here if anyone wanted to take a look: https://github.com/Sam-Sims/minimap2-rs/commit/ad01bbfcb229dba8e6db28081e7f22d1936910f6

This did work for me on my own use case but I didn't have time to properly test it. If that approach seems reasonable I would be happy to open a PR

**Comment by jianshu93** [comment]
hi @rob-p ,

Thanks for debugging my fake minimap2-rs-test, when I started to use it in some large projects, this double free error kept showing up: main(25070,0x16bc4f000) malloc: double free for ptr 0x1783a8000

I am wondering in my case, what would be the best way to get rid of it. I tried to add aligner.idx = None; within each consumer thread without success. 

Thanks,
Best,
Jianshu

**Comment by jianshu93** [comment]
never mind, I've solved it. I also udpated the fake minimap2-rs-test. 

Thanks,

Jianshu

**Comment by jguhlin** [comment]
@rob-p Does this look like it covers your test-case for the double-free errors?

https://github.com/jguhlin/minimap2-rs/blob/bd3eb2247420842ad591b867c9ae4832f5443db0/src/lib.rs#L1932

**Comment by jguhlin** [comment]
Ignore the billion commits. GitButler had an anyeurism

**Comment by rob-p** [comment]
Hi @jguhlin,

  Because of my need to rely on both `minimap2-rs` and `minimap2-sys` directly as dependencies, I can't test this directly in my codebase for `oarfish` (which is why I went the route in the fork of having a separate repo for `minimap2-sys` that gets pulled in).  However, this definitely looks like it covers the cases in question, and should allow us to remove our "hack" from the multithreaded use of the mapper.

**Comment by jguhlin** [comment]
@rob-p Awesome. I made some changes to oarfish to see how things played out. I didn't check the double free but updated the code to work with the new Arc<*mut idx> and added some ergonomic functions

https://github.com/COMBINE-lab/oarfish/pull/36/files

I don't think you need a separate repo, cargo is supposed to recursively search for the specified crate. But, I want to investigate this further. The other ffi crates use this formula, so don't want to go against the idiomatic way just yet... but if it keeps being a problem then yeah, definitely will split it out.

**Comment by rob-p** [comment]
Thanks @jguhlin!

 I didn't know about the recursive search. That fixes the crate import problem. Your changes look good and I tested that I can get rid of the explicit index hack and everything seems to be in working order. From my end, this looks good for a release whenever you are happy with the pending changes. Thanks for all of your work on this!

**Comment by jguhlin** [comment]
@rob-p Just added a typestate builder, to stop people from applying multiple presets at compile time.  

I don't think there will be anything else. Trying to decouple pthreads from minimap2-sys index building, but that doesn't affect anything here.
